redis/src/main/java/org/hibernate/ogm/datastore/redis/AbstractRedisDialect.java

```diff
@@ -316,4 +366,119 @@ protected void storeAssociation(
 			}
 		}
 	}
+
+	/**
+	 * Scan over keys. This method is aware whether the the client is connected to a Redis Cluster.
+	 * If so, then a Redis Cluster scan requires to iterate over master nodes and keep the
+	 * state within a custom {@link KeyScanCursor} instance.
+	 *
+	 * @param cursor
+	 * @param scanArgs
+	 *
+	 * @return
+	 */
+	protected KeyScanCursor<String> scan(KeyScanCursor<String> cursor, ScanArgs scanArgs) {
+
+		if ( !clusterMode ) {
+			return scan( connection, cursor, scanArgs );
+		}
+
+		return clusterScan( cursor, scanArgs );
+	}
+
+	private KeyScanCursor<String> scan(
+			RedisClusterCommands<String, String> commands,
+			KeyScanCursor<String> cursor,
+			ScanArgs scanArgs) {
+		if ( cursor != null ) {
+			return commands.scan( cursor, scanArgs );
+		}
+		return commands.scan( scanArgs );
+	}
+
+	@SuppressWarnings("unchecked")
+	private KeyScanCursor<String> clusterScan(KeyScanCursor<String> cursor, ScanArgs scanArgs) {
+
+		RedisAdvancedClusterCommands<String, String> commands = (RedisAdvancedClusterCommands<String, String>) connection;
+		List<String> nodeIds;
+		String currentNodeId;
+
+		if ( cursor == null ) {
+			Set<RedisClusterNode> masterNodes = commands.masters().asMap().keySet();
+			nodeIds = new ArrayList<>();
+
+			for ( RedisClusterNode masterNode : masterNodes ) {
+				nodeIds.add( masterNode.getNodeId() );
+			}
+
+			if ( nodeIds.isEmpty() ) {
+				return scan( connection, cursor, scanArgs );
+			}
+
+			currentNodeId = nodeIds.get( 0 );
+		}
+		else {
+
+			ClusterwideKeyScanCursor<String> clusterKeyScanCursor = (ClusterwideKeyScanCursor<String>) cursor;
+			nodeIds = clusterKeyScanCursor.nodeIds;
+
+			currentNodeId = getNodeIdForNextScanIteration( nodeIds, clusterKeyScanCursor );
+		}
+
+		RedisClusterCommands<String, String> nodeConnection = commands.getConnection( currentNodeId );
+		KeyScanCursor<String> nodeKeyScanCursor = scan( nodeConnection, cursor, scanArgs );
+
+		return new ClusterwideKeyScanCursor(
+				nodeIds,
+				currentNodeId,
+				nodeKeyScanCursor
+		);
+	}
+
+	private String getNodeIdForNextScanIteration(
+			List<String> nodeIds,
+			ClusterwideKeyScanCursor<String> clusterKeyScanCursor) {
+		if ( clusterKeyScanCursor.isScanOnCurrentNodeFinished() ) {
+
+			int nodeIndex = nodeIds.indexOf( clusterKeyScanCursor.currentNodeId );
+			return nodeIds.get( nodeIndex + 1 );
+		}
+
+		return clusterKeyScanCursor.currentNodeId;
+	}
+
+	/**
+	 * @return {@literal true} if the connected Redis node (default connection) is running in Redis Cluster mode.
+	 */
+	public boolean isClusterMode() {
+		return clusterMode;
+	}
+
+	static class ClusterwideKeyScanCursor<K> extends KeyScanCursor<K> {
```

Comment:
I have to admit I'm a bit confused by the design of these cursors ;)

I would somehow have expected the cursor type(s) to have methods such as hasNext(), next() instead of getKeys(). Also couldn't the `ClusterwideKeyScanCursor` implementation keep all the required state and logic (e.g. `getNodeIdForNextScanIteration()`) internally, so instead of repeatedly invoking `scan()` which returns a new cursor instance each time, there'd be a single invocation returning a single cursor, which then would allow to iterate over all the rows.

WDYT?